codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/aws/SerializeMiddleware.java

```diff
@@ -0,0 +1,223 @@
+    private String serialName;
+
+    public SerializeMiddleware(
+            ProtocolGenerator generator, ProtocolGenerator.GenerationContext ctx, OperationShape operation,
+            GoWriter writer
+    ) {
+        this.generator = generator;
+        this.ctx = ctx;
+        this.operation = operation;
+        this.writer = writer;
+
+        this.input = ctx.getModel().expectShape(operation.getInputShape(), StructureShape.class);
+        this.output = ctx.getModel().expectShape(operation.getOutputShape(), StructureShape.class);
+        this.eventStreamIndex = EventStreamIndex.of(ctx.getModel());
+    }
+
+    @Override
+    public void accept(GoWriter writer) {
+        var name = ProtocolGenerator.getSerializeMiddlewareName(operation.getId(), ctx.getService(),
+                generator.getProtocolName());
+        var middleware = createSerializeStepMiddleware(name, ProtocolUtils.OPERATION_SERIALIZER_MIDDLEWARE_ID);
+//        writer.write(middleware.asWritable(generateOperationStubs(), emptyGoTemplate()));
+    }
+
+    public GoWriter.Writable generateOperationStubs() {
+        this.serialName = "awsAwsjson10_serializeOp" + this.operation.toShapeId().getName();
+
+        return goTemplate("""
+                
+                type $opName:L struct{
+                }
+                
+                func (op *$opName:L) ID() string {
+                    return "OperationSerializer"
+                }
+                
+                $handleSerialize:W
+                
+                """, Map.of(
+                    "opName", this.serialName,
+                    "handleSerialize", generateHandleSerialize()
+        ));
+    }
+
+    private GoWriter.Writable generateHandleSerialize() {
+        return goTemplate(
+                """
+                        
+                        func (op *$opName:L) HandleSerialize (ctx $context:T, in $input:T, next $handler:T) (
+                        out $output:T, metadata $metadata:T, err error) {
+                        
+                            $body:W
+                        
+                            return next.HandleSerialize(ctx, in)
+                        }
+                        
+                        """, Map.of(
+                        "context", SmithyGoDependency.CONTEXT.interfaceSymbol("Context"),
+                        "input", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("SerializeInput"),
+                        "handler", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("SerializeHandler"),
+                        "output", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("SerializeOutput"),
+                        "metadata", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("Metadata"),
+                        "opName", this.serialName,
+                        "body", generateHandleSerializeBody())
+        );
+    }
+
+    private GoWriter.Writable generateHandleSerializeBody() {
+        return goTemplate("""
+                    $route:W
```

Comment:
[don't have to fix]

Investigate re-writing this middleware as one that extends the base `SerializeRequestMiddleware` - that will enable you to just focus on filling out the routing/payload serialize components of the protocol without having to worry about generating the middleware shell, ID method, etc.